contracts/contracts/Tally.sol

```diff
@@ -0,0 +1,248 @@
+
+    Verifier public verifier;
+
+    constructor(Verifier _verifier) {
+        verifier = _verifier;
+    }
+
+    modifier votingPeriodOver(Poll _poll) {
+        (uint256 deployTime, uint256 duration) = _poll
+            .getDeployTimeAndDuration();
+        // Require that the voting period is over
+        uint256 secondsPassed = block.timestamp - deployTime;
+        require(secondsPassed > duration, ERROR_VOTING_PERIOD_NOT_PASSED);
+        _;
+    }
+
+    /*
+     * Pack the batch start index and number of signups into a 100-bit value.
+     */
+    function genTallyVotesPackedVals(
+        uint256 _numSignUps,
+        uint256 _batchStartIndex,
+        uint256 _tallyBatchSize
+    ) public pure returns (uint256) {
+        // TODO: ensure that each value is less than or equal to 2 ** 50
+        uint256 result = (_batchStartIndex / _tallyBatchSize) +
+            (_numSignUps << uint256(50));
+
+        return result;
+    }
+
+    function genTallyVotesPublicInputHash(
+        uint256 _numSignUps,
+        uint256 _batchStartIndex,
+        uint256 _tallyBatchSize,
+        uint256 _newTallyCommitment
+    ) public view returns (uint256) {
+        uint256 packedVals = genTallyVotesPackedVals(
+            _numSignUps,
+            _batchStartIndex,
+            _tallyBatchSize
+        );
+        uint256[] memory input = new uint256[](4);
+        input[0] = packedVals;
+        input[1] = sbCommitment;
+        input[2] = tallyCommitment;
+        input[3] = _newTallyCommitment;
+        uint256 inputHash = sha256Hash(input);
+        return inputHash;
+    }
+
+    // TODO: make sure correct mp address is passed or change to private function
+    // TODO: reuse tally.sol for multiple polls
+    function updateSbCommitment(MessageProcessor _mp) public {
+        // Require that all messages have been processed
+        require(_mp.processingComplete(), ERROR_PROCESSING_NOT_COMPLETE);
+        if (sbCommitment == 0) {
+            sbCommitment = _mp.sbCommitment();
+        }
+    }
+
+    function tallyVotes(
+        Poll _poll,
+        MessageProcessor _mp,
+        uint256 _newTallyCommitment,
+        uint256[8] memory _proof
+    ) public onlyOwner votingPeriodOver(_poll) {
+        updateSbCommitment(_mp);
+
+        (, uint256 tallyBatchSize, ) = _poll.batchSizes();
+        uint256 batchStartIndex = tallyBatchNum * tallyBatchSize;
+        (uint256 numSignUps, ) = _poll.numSignUpsAndMessages();
+
+        // Require that there are untalied ballots left
+        require(batchStartIndex <= numSignUps, ERROR_ALL_BALLOTS_TALLIED);
+
+        bool isValid = verifyTallyProof(
+            _poll,
+            _proof,
+            numSignUps,
+            batchStartIndex,
+            tallyBatchSize,
+            _newTallyCommitment
+        );
+        require(isValid, ERROR_INVALID_TALLY_VOTES_PROOF);
+
+        // Update the tally commitment and the tally batch num
+        tallyCommitment = _newTallyCommitment;
```

Comment:
Move these two lines up at the top (checks effect interaction pattern). You can store `tallyBatchNum` in a local variable and use that for calculations:

```js
uint256 _tallyBatchNum = tallyBatchNum;
unchecked {
    tallyBatchNum++;
}
```

Reply:
After this modification, it will cause stack too deep in this function. Since onlyOwner can call this function, it should be okay. What do you think?

Reply:
Right ok, it should  be fine to move only the commitment update on top and keep the counter increased down given that it's protected by onlyOwner. We can always try and refactor it later.

Reply:
Yeah, that should work.